File: Python/VigenereCipher/VigenereCipher.py

ALPHABET = ' ABCDEFGHIJKLMNOPQRSTUVWXYZ'

def vigenere_encrypt(plain_text, key):
    plain_text = plain_text.upper()
    key = key.upper()

    cipher_text = ''
    key_index = 0

    for character in plain_text:
        index = (ALPHABET.find(character) + ALPHABET.find(key[key_index])) % len(ALPHABET)
        cipher_text = cipher_text + ALPHABET[index]
        key_index = key_index + 1

        if key_index == len(key):
            key_index = 0

    return cipher_text


def vigenere_decrypt(cipher_text, key):
    cipher_text = cipher_text.upper()
    key = key.upper()

    plain_text = ''
    key_index = 0

    for character in cipher_text:
        index = (ALPHABET.find(character) - ALPHABET.find(key[key_index])) % len(ALPHABET)
        plain_text = plain_text + ALPHABET[index]
        key_index = key_index + 1

        if key_index == len(key):
            key_index = 0

    return plain_text


if __name__ == '__main__':
    text = 'CRYPTOGRAPHY IS QUITE IMPORTANT IN CRYPTOCURRENCIES'
    print(vigenere_encrypt(text, 'ANIMAL'))
    cipher_text = vigenere_encrypt(text, 'ANIMAL')
    print(vigenere_decrypt(cipher_text, 'ANIMAL'))

File: Java/VigenereCipher/VigenereCipher.java
package Java.VigenereCipher;

public class VigenereCipher {

    private static final String ALPHABET = " ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    public String encrypt(String plainText, String key) {
        plainText = plainText.toUpperCase();
        key = key.toUpperCase();

        StringBuilder cipherText = new StringBuilder();
        int keyIndex = 0;

        for (int i = 0; i < plainText.length(); i++) {
            char c = plainText.charAt(i);
            int index = Math.floorMod(ALPHABET.indexOf(c) + ALPHABET.indexOf(key.charAt(keyIndex)), ALPHABET.length());

            cipherText.append(ALPHABET.charAt(index));
            keyIndex++;

            if (keyIndex == key.length()) {
                keyIndex = 0;
            }
        }

        return cipherText.toString();
    }

    public String decrypt(String cipherText, String key) {
        cipherText = cipherText.toUpperCase();
        key = key.toUpperCase();

        StringBuilder plainText = new StringBuilder();
        int keyIndex = 0;

        for (int i = 0; i < cipherText.length(); i++) {
            char c = cipherText.charAt(i);
            int index = Math.floorMod(ALPHABET.indexOf(c) - ALPHABET.indexOf(key.charAt(keyIndex)), ALPHABET.length());

            plainText.append(ALPHABET.charAt(index));
            keyIndex++;

            if (keyIndex == key.length()) {
                keyIndex = 0;
            }
        }

        return plainText.toString();
    }
}
